lib/semantics/check-declarations.cc

```diff
@@ -352,6 +281,251 @@ void CheckHelper::CheckValue(
+                interface->attrs().test(Attr::ELEMENTAL)))) {
+      // There's no explicit constraint or "shall" that we can find in the
+      // standard for this check, but it seems to be implied in multiple
+      // sites, and ELEMENTAL non-intrinsic actual arguments *are*
+      // explicitly forbidden.  But we allow "PROCEDURE(SIN)::dummy"
+      // because it is explicitly legal to *pass* the specific intrinsic
+      // function SIN as an actual argument.
+      messages_.Say("A dummy procedure may not be ELEMENTAL"_err_en_US);
+    }
+  }
+}
+
+void CheckHelper::CheckDerivedType(
+    const Symbol &symbol, const DerivedTypeDetails &details) {
+  CHECK(symbol.scope());
+  CHECK(symbol.scope()->symbol() == &symbol);
+  CHECK(symbol.scope()->IsDerivedType());
+  if (symbol.attrs().test(Attr::ABSTRACT) &&
+      (symbol.attrs().test(Attr::BIND_C) || details.sequence())) {
+    messages_.Say("An ABSTRACT derived type must be extensible"_err_en_US);
+  }
+  if (const DeclTypeSpec * parent{FindParentTypeSpec(symbol)}) {
+    const DerivedTypeSpec *parentDerived{parent->AsDerived()};
+    if (!IsExtensibleType(parentDerived)) {
+      messages_.Say("The parent type is not extensible"_err_en_US);
+    }
+    if (!symbol.attrs().test(Attr::ABSTRACT) && parentDerived &&
+        parentDerived->typeSymbol().attrs().test(Attr::ABSTRACT)) {
+      ScopeComponentIterator components{*parentDerived};
+      for (const Symbol &component : components) {
+        if (component.attrs().test(Attr::DEFERRED)) {
+          if (symbol.scope()->FindComponent(component.name()) == &component) {
+            SayWithDeclaration(component,
+                "Non-ABSTRACT extension of ABSTRACT derived type '%s' lacks a binding for DEFERRED procedure '%s'"_err_en_US,
+                parentDerived->typeSymbol().name(), component.name());
+          }
+        }
+      }
+    }
+  }
+}
+
+void CheckHelper::CheckGeneric(
+    const Symbol &symbol, const GenericDetails &details) {
+  const SymbolVector &specifics{details.specificProcs()};
+  const auto &bindingNames{details.bindingNames()};
+  std::optional<std::vector<Procedure>> procs{Characterize(specifics)};
+  if (!procs) {
+    return;
+  }
+  bool ok{true};
+  if (details.kind().IsAssignment()) {
+    for (std::size_t i{0}; i < specifics.size(); ++i) {
+      auto save{messages_.SetLocation(bindingNames[i])};
+      ok &= CheckDefinedAssignment(specifics[i], (*procs)[i]);
+    }
+  }
+  // TODO: check defined operators too
+  if (ok) {
+    CheckSpecificsAreDistinguishable(symbol, details, *procs);
+  }
+}
+
+// Check that the specifics of this generic are distinguishable from each other
+void CheckHelper::CheckSpecificsAreDistinguishable(const Symbol &generic,
+    const GenericDetails &details, const std::vector<Procedure> &procs) {
+  const SymbolVector &specifics{details.specificProcs()};
+  auto count{specifics.size()};
+  if (specifics.size() < 2) {
```

Comment:
Should you just use "count" rather than "specifics.size()"?

Reply:
Right, fixed.